app/components/Taps/TapConfiguration/S3/index.js

```diff
@@ -0,0 +1,350 @@
+
+type Props = {
+  tapsStore: {
+    'tap-s3-csv': {
+      fieldValues: {
+        bucket: string,
+        start_date: string,
+        tables: string
+      }
+    }
+  },
+  updateTapField: (tap: string, field: string, value: string | number) => void
+};
+type State = {
+  bucket: '',
+  start_date: '',
+  tables: [{ table_name: '', key_properties: [], search_pattern: '' }],
+  tags: []
+};
+
+export default class S3 extends Component<Props, State> {
+  constructor(props) {
+    super(props);
+    this.state = {
+      bucket: '',
+      start_date: '',
+      tables: [{ table_name: '', key_properties: [], search_pattern: '' }],
+      tags: []
+    };
+  }
+
+  componentWillMount = () => {
+    const tagObj = {};
+    const tagsArr = [];
+    const { tables } = this.props.tapsStore['tap-s3-csv'].fieldValues;
+    const tableValue = tables
+      ? JSON.parse(tables)
+      : [{ table_name: '', key_properties: [], search_pattern: '' }];
+    tableValue.forEach((key) => {
+      const keyProperties = key.key_properties;
+      if (keyProperties.length !== 0) {
+        tagObj.id = keyProperties;
+        tagObj.text = keyProperties;
+        tagsArr.push(tagObj);
+      }
+    });
+    this.setState({
+      tables: tableValue,
+      tags: tagsArr
+    });
+  };
+
+  validate = (field: string, value: string) => {
+    if (value) {
+      this.setState({ [field]: { valid: true } });
+    } else {
+      this.setState({ [field]: { invalid: true } });
+    }
+  };
+
+  toISODateString = (date: Date) => {
+    const pad = (number) => (number < 10 ? `0${number}` : number);
+
+    return `${date.getUTCFullYear()}-${pad(date.getUTCMonth() + 1)}-${pad(
+      date.getUTCDate()
+    )}T${pad(date.getUTCHours())}:${pad(date.getUTCMinutes())}:${pad(
+      date.getUTCSeconds()
+    )}Z`;
+  };
+
+  handleChange = (e: SyntheticEvent<HTMLButtonElement>) => {
+    const { name } = e.currentTarget;
+    let { value } = e.currentTarget;
+
+    if (name === 'start_date') {
+      value = this.toISODateString(new Date(value));
+    }
+
+    this.props.updateTapField('tap-s3-csv', name, value);
+  };
+
+  formatDate = (ISODate: string) => {
+    const date = new Date(ISODate);
+    const year = date.getFullYear();
+    const month = date.getMonth() + 1;
+    const day = date.getDate();
+
+    return `${year}-${month < 10 ? `0${month}` : month}-${
+      day < 10 ? `0${day}` : day
+    }`;
+  };
+
+  updateTable = (newTable) => {
+    this.setState({ tables: newTable }, () => {
+      const tableValue = JSON.stringify(this.state.tables);
+      this.props.updateTapField('tap-s3-csv', 'tables', tableValue);
+    });
+  };
+
+  handleAddTable = () => {
+    this.setState({
+      tables: this.state.tables.concat([
+        {
+          table_name: '',
+          key_properties: [],
+          search_pattern: ''
+        }
+      ])
+    });
+  };
+
+  handleRemoveTable = (idx) => () => {
+    const newTable = this.state.tables.filter((s, sidx) => idx !== sidx);
```

Comment:
I think state changes should always be performed using `this.setState()`

Reply:
`updateTable` method is calling `this.setState()` to update the state in here.